### PR TITLE
posture: add get serial support for Windows/Linux

### PR DESCRIPTION
**Pull request by kradalby**
This commit adds support for getting serial numbers from SMBIOS on Windows/Linux (and BSD) using go-smbios.

This means that it will add a new dependency, [go-smbios](https://github.com/digitalocean/go-smbios/tree/master).
go-smbios is a small library that allows go to read the smbios/dmi table on Windows/Linux/BSD which contains various information like serial numbers.

While reviewing this PR, please also have a look at the new dependency, it was chosen because

- It solves the problem for Windows and Linux
- It was made by well-known Digital Ocean and [Matt Layher](https://github.com/mdlayher)
- It has not been changed in a long time, low amount of issues, indicating stability
- It has [decent test coverage](https://github.com/search?q=repo%3Adigitalocean%2Fgo-smbios%20_test.go&type=code)
- It has no external dependencies (except go-cmp for tests)

Updates #5902

Ubuntu in a VM:
image

Windows 11 ARM in a VM:
image

### CODE REVIEW

File: posture/serialnumber_notmacos.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		return "", errors.New("offset does not exist in smbios structure")
+	}
+
+	str := s.Strings[index-1]
+	trimmed := strings.TrimSpace(str)
+
+	// Convert to lowercase to address multiple formats:
+	//   - "To Be Filled By O.E.M."
+	//   - "To be filled by O.E.M."
+	if strings.ToLower(trimmed) == "to be filled by o.e.m." {
```
Comment (kradalby) [inline posture/serialnumber_notmacos.go:44]: This could be a prefix check instead, "to be filled", maybe something else is in the wild out there.

**Review by dblohm7 — APPROVED** [review]
Thank you for selecting a package that does not include a deep tree of third-party dependencies! That package seems okay. I have a minor nit with the way they made the API call but we've already mitigated that using other mechanisms.

LGTM provided comments are considered and other reviewers approve.

File: posture/serialnumber_notmacos.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+	str := s.Strings[index-1]
+	trimmed := strings.TrimSpace(str)
+
+	// Convert to lowercase to address multiple formats:
+	//   - "To Be Filled By O.E.M."
+	//   - "To be filled by O.E.M."
+	if strings.EqualFold(trimmed, "to be filled by o.e.m.") {
+		return "", errors.New("data is not provided by OEM")
+	}
```
Comment (knyar) [inline posture/serialnumber_notmacos.go:46]: I was thinking that we will just filter such bogus values on control side, but we can definitely do a first pass in the client as well. I don't think it should be an error though, as we do expect some machines to not have a serial number we can collect.

File: posture/serialnumber_notmacos.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/digitalocean/go-smbios/smbios"
+	"tailscale.com/util/multierr"
+)
+
+// getByteFromSmbiosStructure retrieves a 8-bit unsigned integer at the given specOffset.
+func getByteFromSmbiosStructure(s *smbios.Structure, specOffset int) uint8 {
+	// the `Formatted` byte slice is missing the first 4 bytes of the structure that are stripped out as header info.
+	// so we need to subtract 4 from the offset mentioned in the SMBIOS documentation to get the right value.
+	index := specOffset - 4
+	if index >= len(s.Formatted) {
```
Comment (knyar) [inline posture/serialnumber_notmacos.go:22]: Do you want to add `|| index < 0` just in case someone passes specOffset smaller than 4?

**Review by knyar — APPROVED** [review]

File: posture/serialnumber_notmacos.go (hunk 2 above)
Comment (kradalby) [reply posture/serialnumber_notmacos.go:46]: Let's filter it out on the control side.

**Review by knyar — APPROVED** [review]

File: posture/serialnumber_notmacos.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+			serials = append(serials, serial)
+		}
+	}
+
+	// if there were no serial numbers, check if any errors were
+	// returned and combine them.
+	if len(serials) == 0 && len(errs) > 0 {
+		return nil, multierr.New(errs...)
+	}
+
```
Comment (knyar) [inline posture/serialnumber_notmacos.go:136]: I understand why we might want to ignore errors if were able to collect at least one serial number, but perhaps we should at least print those errors here.

File: posture/serialnumber_notmacos_test.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Build on Windows, Linux and *BSD
+
+//go:build !darwin && !android && !js && !plan9 && !solaris
```
Comment (knyar) [inline posture/serialnumber_notmacos_test.go:6]: Let's match the other file?

```suggestion
//go:build windows || (linux && !android) || freebsd || openbsd || dragonfly || netbsd
```

File: posture/serialnumber_test.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+
+package posture
+
+import (
+	"testing"
+)
+
+func TestGetSerialNumber(t *testing.T) {
+	// ensure GetSerialNumbers is implemented
```
Comment (knyar) [inline posture/serialnumber_test.go:11]: 👍

**Review by knyar — APPROVED** [review]

**Pull request merged**